### PR TITLE
Filter problems by difficulty

### PR DESCRIPTION
**Pull request by SaySayo**
Filtered problems by difficulty in the exercises page OCaml.org 

Fixes the second task in #890.

### CODE REVIEW

**Review by sabine — COMMENTED** [review]
I think we need to remove the `overflow-hidden` on the div in line 40 because the outline of the select is cut off.

It's probably better to write "Any difficulty", instead of "Difficulty" in the select option because it makes more sense when the select is opened.

Overall LGTM. Just these two changes and it's mergeable.

**Review by sabine — APPROVED** [review]

File: src/ocamlorg_web/lib/handler.ml [hunk 1]
```diff
@@ -278,7 +278,32 @@ let best_practices _req =
...
 
-let problems _req = Dream.html (Ocamlorg_frontend.problems Ood.Problem.all)
+let problems req =
+  let all_problems = Ood.Problem.all in
+  let difficulty_level = Dream.query req "difficulty_level" in
+  let difficulty_of_string = function
+    | "beginner" -> `Beginner
+    | "intermediate" -> `Intermediate
+    | "advanced" -> `Advanced
+    | _ -> failwith "Invalid difficulty string"
```
Comment (cuihtlauac) [inline src/ocamlorg_web/lib/handler.ml]:
```suggestion
    | "beginner" -> ( = ) `Beginner
    | "intermediate" -> ( = ) `Intermediate
    | "advanced" -> ( = ) `Advanced
    | _ -> Fun.const true
```

File: src/ocamlorg_web/lib/handler.ml [hunk 2]
```diff
@@ -278,7 +278,32 @@ let best_practices _req =
...
+        problem.difficulty = difficulty_of_string difficulty)
+      problems
+  in
+  let filtered_problems =
+    match difficulty_level with
+    | Some "All" -> Ood.Problem.all
+    | Some "beginner" -> filter_by_difficulty all_problems "beginner"
+    | Some "intermediate" -> filter_by_difficulty all_problems "intermediate"
+    | Some "advanced" -> filter_by_difficulty all_problems "advanced"
+    | None | _ -> Ood.Problem.all
```
Comment (cuihtlauac) [inline src/ocamlorg_web/lib/handler.ml]:
```suggestion
  let filtered_problems = List.filter (by_difficulty difficulty_level) all_problems in
```

File: src/ocamlorg_web/lib/handler.ml [hunk 3]
```diff
@@ -278,7 +278,32 @@ let best_practices _req =
...
+    | "intermediate" -> `Intermediate
+    | "advanced" -> `Advanced
+    | _ -> failwith "Invalid difficulty string"
+  in
+  let filter_by_difficulty (problems : Ood.Problem.t list) (difficulty : string)
+      =
+    List.filter
+      (fun (problem : Ood.Problem.t) ->
+        problem.difficulty = difficulty_of_string difficulty)
+      problems
```
Comment (cuihtlauac) [inline src/ocamlorg_web/lib/handler.ml]:
```suggestion
  let by_difficulty level (problem : Ood.Problem.t) =
    match level with
    | Some difficulty -> compare_difficulty difficulty problem.difficulty
    | _ -> true
```

File: src/ocamlorg_web/lib/handler.ml [hunk 4]
```diff
@@ -278,7 +278,32 @@ let best_practices _req =
   let tutorials = Ood.Tutorial.all in
   Dream.html (Ocamlorg_frontend.best_practices ~tutorials Ood.Workflow.all)
 
-let problems _req = Dream.html (Ocamlorg_frontend.problems Ood.Problem.all)
+let problems req =
+  let all_problems = Ood.Problem.all in
+  let difficulty_level = Dream.query req "difficulty_level" in
+  let difficulty_of_string = function
```
Comment (cuihtlauac) [inline src/ocamlorg_web/lib/handler.ml]:
```suggestion
  let compare_difficulty = function
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

**Review by cuihtlauac — APPROVED** [review]

**Pull request merged**